File: bolao/shared/scripts/test_notification_repository.mjs
#!/usr/bin/env node
// test_notification_repository.mjs — real tests for MemoryNotificationRepository and
// FileNotificationRepository against the shared NotificationRepository contract (football-
// hardening readiness follow-up, items 2/5). SupabaseNotificationRepository is code-complete
// but NOT tested here — see docs/bolao/FOOTBALL_HARDENING_SUPABASE_TEST_EXECUTION.md for why
// (no test Supabase project available) and its own honest NÃO EXECUTADO status.
//
// Run: node bolao/shared/scripts/test_notification_repository.mjs

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  MemoryNotificationRepository, FileNotificationRepository,
  toCanonical, fromCanonical, buildIdempotencyKey, JOB_STATUS, SCHEMA_VERSION,
} from "./notification_repository.mjs";
import { makeFakeClock } from "./fake_clock.mjs";

const results = [];
function check(name, pass, detail) {
  results.push({ name, pass });
  console.log(`${pass ? "✓" : "✗"} ${name}`);
  if (!pass && detail !== undefined) console.error(`    ${JSON.stringify(detail)}`);
}

async function runSuite(label, makeRepo) {
  // ── Same battery of real behavior checks, run against EVERY adapter — Copa/BR/CDB must see
  // identical behavior regardless of backend. ────────────────────────────────────────────────
  {
    const repo = makeRepo();
    const { event, created } = await repo.createEvent({ poolId: "cdb2026", entityType: "tie", entityId: "oitavas:tie-1", eventType: "final_confirmed", eventVersion: 1, payloadSnapshot: { homeScore: 1, awayScore: 1 } });
    check(`${label}: createEvent() creates on first call`, created === true, event);
    const dup = await repo.createEvent({ poolId: "cdb2026", entityType: "tie", entityId: "oitavas:tie-1", eventType: "final_confirmed", eventVersion: 1, payloadSnapshot: { homeScore: 1, awayScore: 1 } });
    check(`${label}: createEvent() is idempotent (same pool/entity/type/version -> created:false)`, dup.created === false && dup.event.eventId === event.eventId, dup);

    const key1 = buildIdempotencyKey("cdb2026", "oitavas:tie-1", "alfa@example.test", 1);
    const key2 = buildIdempotencyKey("cdb2026", "oitavas:tie-1", "beta@example.test", 1);
    const { jobs, createdCount } = await repo.enqueueJobs(event.eventId, [
      { poolId: "cdb2026", recipient: "alfa@example.test", payloadSnapshot: { x: 1 }, idempotencyKey: key1 },
      { poolId: "cdb2026", recipient: "beta@example.test", payloadSnapshot: { x: 1 }, idempotencyKey: key2 },
    ]);
    check(`${label}: enqueueJobs() creates 2 jobs`, createdCount === 2 && jobs.length === 2, { createdCount, jobsLen: jobs.length });

    const dupJobs = await repo.enqueueJobs(event.eventId, [
      { poolId: "cdb2026", recipient: "alfa@example.test", payloadSnapshot: { x: 1 }, idempotencyKey: key1 },
    ]);
    check(`${label}: enqueueJobs() is idempotent (same idempotencyKey -> 0 new)`, dupJobs.createdCount === 0, dupJobs);

    const claimed = await repo.claimPendingJobs("cdb2026", 50, "test-worker");
    check(`${label}: claimPendingJobs() claims exactly 2`, claimed.length === 2 && claimed.every((j) => j.status === JOB_STATUS.PROCESSING), claimed.map((j) => j.status));

    const claimedAgain = await repo.claimPendingJobs("cdb2026", 50, "test-worker-2");
    check(`${label}: claimPendingJobs() a second time claims ZERO (already processing)`, claimedAgain.length === 0, claimedAgain);

    const sent = await repo.markSent(claimed[0].jobId, { providerMessageId: "provider-msg-1" });
    check(`${label}: markSent() transitions to sent with providerMessageId`, sent.status === JOB_STATUS.SENT && sent.providerMessageId === "provider-msg-1", sent);

    const failed = await repo.markRetryableFailure(claimed[1].jobId, { error: "simulated provider outage" });
    check(`${label}: markRetryableFailure() transitions to failed_retryable with error recorded`, failed.status === JOB_STATUS.FAILED_RETRYABLE && failed.lastError === "simulated provider outage", failed);
  }

  // ── Stuck job recovery ──────────────────────────────────────────────────────────────────
  {
    const clock = makeFakeClock();
    const repo = makeRepo(clock);
    const { event } = await repo.createEvent({ poolId: "br2026", entityType: "round_batch", entityId: "round:g1,g2", eventType: "final_confirmed", eventVersion: 1, payloadSnapshot: {} });
    await repo.enqueueJobs(event.eventId, [{ poolId: "br2026", recipient: "x@example.test", payloadSnapshot: {}, idempotencyKey: buildIdempotencyKey("br2026", "round:g1,g2", "x@example.test", 1) }]);
    await repo.claimPendingJobs("br2026", 10, "crashed-worker");
    clock.advanceMinutes(10);
    const recovered = await repo.releaseStuckJobs(5 * 60 * 1000);
    check(`${label}: releaseStuckJobs() recovers exactly 1 stuck job`, recovered === 1, recovered);
  }

  // ── Missing notifications ───────────────────────────────────────────────────────────────
  {
    const repo = makeRepo();
    const { event } = await repo.createEvent({ poolId: "copa2026", entityType: "match", entityId: "M95", eventType: "final_confirmed", eventVersion: 1, payloadSnapshot: {} });
    const missing = await repo.findMissingNotifications("copa2026");
    check(`${label}: findMissingNotifications() finds an event with zero jobs`, missing.some((e) => e.eventId === event.eventId), missing);
    await repo.enqueueJobs(event.eventId, [{ poolId: "copa2026", recipient: "y@example.test", payloadSnapshot: {}, idempotencyKey: buildIdempotencyKey("copa2026", "M95", "y@example.test", 1) }]);
    const missingAfter = await repo.findMissingNotifications("copa2026");
    check(`${label}: findMissingNotifications() no longer lists it once jobs exist`, !missingAfter.some((e) => e.eventId === event.eventId), missingAfter);
  }
}

const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "notif-repo-test-"));

await runSuite("MemoryNotificationRepository", (clock) => new MemoryNotificationRepository(clock || makeFakeClock()));
await runSuite("FileNotificationRepository", (clock) => new FileNotificationRepository(clock || makeFakeClock(), {
  eventsPath: path.join(fs.mkdtempSync(path.join(tmpDir, "run-")), "events.json"),
  jobsPath: path.join(fs.mkdtempSync(path.join(tmpDir, "run-")), "jobs.json"),
}));

// ── Canonical schema compatibility (item 5) ────────────────────────────────────────────────
{
  const oldJob = {
    schemaVersion: 1, jobId: "job_1", app: "cdb2026", matchId: "oitavas:tie-1", recipient: "a@x.test",
    resultVersion: 1, payloadSnapshot: { x: 1 }, idempotencyKey: "cdb2026:oitavas:tie-1:a@x.test:v1",
    status: "sent", attemptCount: 1, maxAttempts: 5, processingStartedAt: null, lastAttemptAt: "2026-01-01T00:00:00Z",
    sentAt: "2026-01-01T00:00:00Z", providerMessageId: "pm-1", lastError: null, createdAt: "2026-01-01T00:00:00Z",
  };
  const canonical = toCanonical(oldJob);
  check("toCanonical(): maps app->poolId, matchId->entityId, resultVersion->eventVersion", canonical.poolId === "cdb2026" && canonical.entityId === "oitavas:tie-1" && canonical.eventVersion === 1, canonical);
  check("toCanonical(): schemaVersion bumped to current SCHEMA_VERSION", canonical.schemaVersion === SCHEMA_VERSION, canonical.schemaVersion);
  check("toCanonical(): preserves providerMessageId", canonical.providerMessageId === "pm-1", canonical.providerMessageId);

  const roundTripped = fromCanonical(canonical);
  check("fromCanonical(): round-trips back to the old field names", roundTripped.app === "cdb2026" && roundTripped.matchId === "oitavas:tie-1" && roundTripped.resultVersion === 1, roundTripped);
  check("fromCanonical(): round-trips status sent (no failed_retryable/failed_permanent split in v1)", roundTripped.status === "sent", roundTripped.status);
}

fs.rmSync(tmpDir, { recursive: true, force: true });

const pass = results.filter((r) => r.pass).length;
console.log(`\n${pass}/${results.length} checks passed.`);
if (pass !== results.length) {
  console.error(`✗ ${results.length - pass} check(s) FAILED`);
  process.exit(1);
}
console.log("✓ ALL NOTIFICATION REPOSITORY CHECKS PASSED");
process.exit(0);

File: bolao/shared/scripts/test_notification_repository.py
"""
test_notification_repository.py — real tests for MemoryNotificationRepository and
FileNotificationRepository (Python side), mirroring test_notification_repository.mjs exactly so
Node and Python are proven to behave identically against the shared contract (football-hardening
readiness follow-up, items 2/5).

Run: python3 bolao/shared/scripts/test_notification_repository.py
"""
import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from notification_repository import (  # noqa: E402
    MemoryNotificationRepository, FileNotificationRepository,
    to_canonical, from_canonical, build_idempotency_key, JOB_STATUS, SCHEMA_VERSION,
)


def make_repo_suite(repo_factory, label):
    class Suite(unittest.TestCase):
        def test_create_event_and_enqueue_and_claim_and_mark(self):
            repo = repo_factory()
            event, created = repo.create_event({"poolId": "cdb2026", "entityType": "tie", "entityId": "oitavas:tie-1", "eventType": "final_confirmed", "eventVersion": 1, "payloadSnapshot": {"homeScore": 1, "awayScore": 1}})
            self.assertTrue(created)
            dup_event, dup_created = repo.create_event({"poolId": "cdb2026", "entityType": "tie", "entityId": "oitavas:tie-1", "eventType": "final_confirmed", "eventVersion": 1, "payloadSnapshot": {"homeScore": 1, "awayScore": 1}})
            self.assertFalse(dup_created)
            self.assertEqual(dup_event["eventId"], event["eventId"])

            key1 = build_idempotency_key("cdb2026", "oitavas:tie-1", "alfa@example.test", 1)
            key2 = build_idempotency_key("cdb2026", "oitavas:tie-1", "beta@example.test", 1)
            jobs, created_count = repo.enqueue_jobs(event["eventId"], [
                {"poolId": "cdb2026", "recipient": "alfa@example.test", "payloadSnapshot": {"x": 1}, "idempotencyKey": key1},
                {"poolId": "cdb2026", "recipient": "beta@example.test", "payloadSnapshot": {"x": 1}, "idempotencyKey": key2},
            ])
            self.assertEqual(created_count, 2)
            self.assertEqual(len(jobs), 2)
            self.assertEqual(jobs[0]["entityId"], "oitavas:tie-1", "job must carry denormalized entityId")
            self.assertEqual(jobs[0]["eventVersion"], 1, "job must carry denormalized eventVersion")

            dup_jobs, dup_count = repo.enqueue_jobs(event["eventId"], [
                {"poolId": "cdb2026", "recipient": "alfa@example.test", "payloadSnapshot": {"x": 1}, "idempotencyKey": key1},
            ])
            self.assertEqual(dup_count, 0, "enqueueJobs must be idempotent by idempotencyKey")

            claimed = repo.claim_pending_jobs("cdb2026", 50, "test-worker")
            self.assertEqual(len(claimed), 2)
            self.assertTrue(all(j["status"] == JOB_STATUS["PROCESSING"] for j in claimed))

            claimed_again = repo.claim_pending_jobs("cdb2026", 50, "test-worker-2")
            self.assertEqual(len(claimed_again), 0, "a second claim must see zero already-claimed jobs")

            sent = repo.mark_sent(claimed[0]["jobId"], provider_message_id="provider-msg-1")
            self.assertEqual(sent["status"], JOB_STATUS["SENT"])
            self.assertEqual(sent["providerMessageId"], "provider-msg-1")

            failed = repo.mark_retryable_failure(claimed[1]["jobId"], error="simulated provider outage")
            self.assertEqual(failed["status"], JOB_STATUS["FAILED_RETRYABLE"])
            self.assertEqual(failed["lastError"], "simulated provider outage")

        def test_release_stuck_jobs(self):
            repo = repo_factory()
            event, _ = repo.create_event({"poolId": "br2026", "entityType": "round_batch", "entityId": "round:g1,g2", "eventType": "final_confirmed", "eventVersion": 1, "payloadSnapshot": {}})
            repo.enqueue_jobs(event["eventId"], [{"poolId": "br2026", "recipient": "x@example.test", "payloadSnapshot": {}, "idempotencyKey": build_idempotency_key("br2026", "round:g1,g2", "x@example.test", 1)}])
            repo.claim_pending_jobs("br2026", 10, "crashed-worker")
            # Fake clock via now_ms override — 10 real minutes later.
            import time
            future_ms = int(time.time() * 1000) + 10 * 60 * 1000
            recovered = repo.release_stuck_jobs(5 * 60 * 1000, now_ms=future_ms)
            self.assertEqual(recovered, 1)

        def test_find_missing_notifications(self):
            repo = repo_factory()
            event, _ = repo.create_event({"poolId": "copa2026", "entityType": "match", "entityId": "M95", "eventType": "final_confirmed", "eventVersion": 1, "payloadSnapshot": {}})
            missing = repo.find_missing_notifications("copa2026")
            self.assertTrue(any(e["eventId"] == event["eventId"] for e in missing))
            repo.enqueue_jobs(event["eventId"], [{"poolId": "copa2026", "recipient": "y@example.test", "payloadSnapshot": {}, "idempotencyKey": build_idempotency_key("copa2026", "M95", "y@example.test", 1)}])
            missing_after = repo.find_missing_notifications("copa2026")
            self.assertFalse(any(e["eventId"] == event["eventId"] for e in missing_after))

    Suite.__name__ = f"{label}Tests"
    Suite.__qualname__ = Suite.__name__
    return Suite


MemoryTests = make_repo_suite(lambda: MemoryNotificationRepository(), "Memory")


class _FileRepoFactory:
    def __init__(self):
        self._tmpdir = tempfile.TemporaryDirectory()

    def __call__(self):
        d = tempfile.mkdtemp(dir=self._tmpdir.name)
        return FileNotificationRepository(events_path=str(Path(d) / "events.json"), jobs_path=str(Path(d) / "jobs.json"))


_file_factory = _FileRepoFactory()
FileTests = make_repo_suite(_file_factory, "File")


class CanonicalSchemaTests(unittest.TestCase):
    def test_to_canonical_maps_old_names(self):
        old_job = {
            "schemaVersion": 1, "jobId": "job_1", "app": "cdb2026", "matchId": "oitavas:tie-1",
            "recipient": "a@x.test", "resultVersion": 1, "payloadSnapshot": {"x": 1},
            "idempotencyKey": "cdb2026:oitavas:tie-1:a@x.test:v1", "status": "sent", "attemptCount": 1,
            "maxAttempts": 5, "processingStartedAt": None, "lastAttemptAt": "2026-01-01T00:00:00Z",
            "sentAt": "2026-01-01T00:00:00Z", "providerMessageId": "pm-1", "lastError": None,
            "createdAt": "2026-01-01T00:00:00Z",
        }
        canonical = to_canonical(old_job)
        self.assertEqual(canonical["poolId"], "cdb2026")
        self.assertEqual(canonical["entityId"], "oitavas:tie-1")
        self.assertEqual(canonical["eventVersion"], 1)
        self.assertEqual(canonical["schemaVersion"], SCHEMA_VERSION)
        self.assertEqual(canonical["providerMessageId"], "pm-1")

    def test_from_canonical_round_trips(self):
        old_job = {"schemaVersion": 1, "jobId": "job_1", "app": "cdb2026", "matchId": "oitavas:tie-1", "recipient": "a@x.test", "resultVersion": 1, "payloadSnapshot": {"x": 1}, "idempotencyKey": "k1", "status": "sent", "attemptCount": 1, "maxAttempts": 5, "processingStartedAt": None, "lastAttemptAt": None, "sentAt": None, "providerMessageId": None, "lastError": None, "createdAt": None}
        canonical = to_canonical(old_job)
        roundtripped = from_canonical(canonical)
        self.assertEqual(roundtripped["app"], "cdb2026")
        self.assertEqual(roundtripped["matchId"], "oitavas:tie-1")
        self.assertEqual(roundtripped["resultVersion"], 1)
        self.assertEqual(roundtripped["status"], "sent")


if __name__ == "__main__":
    unittest.main(verbosity=2)
